### ISSUE TITLE
Using ByteScheduler is not as fast as Ring-allreduce

### ISSUE THREAD

**Issue by Richie-yan** [bytedance/byteps#339]
Hi, I applied ByteScheduler to Horovod+Pytorch, and ran pytorch_horovod_benchmark.py to compare the speed with Horovod+Pytorch without ByteScheduler, and found:

Restnet50 models:
ByteScheduler+Horovod+Pytorch: 180-200 img/sec per GPU
Horovod+Pytorch: 254 img/sec per GPU

The speed is much worse. Remember that ByteScheduler mentioned in the paper can increase by 11%-15%. Could you tell me where the parameters are adjusted incorrectly?
I try to adjust BYTESCHEDULER_CREDIT and BYTESCHEDULER_PARTITION
It is found that the speed improvement is not great. Is there a recommended configuration size for these two parameters?

**Comment by bobzhuyb** [comment]
+ @pengyanghua 

Can you tell us your setup? like, how many GPUs/machines, batch size, etc.? ResNet is pretty light in communication, so you won't get significant improvement anyways.

**Comment by Richie-yan** [comment]
Hi @bobzhuyb 
I ran 8 GPUS in a PCIE machine, batch_size is: 32
I am curious to ask that the resnet50 income in the paper is 11%-15%.
In addition, when I was running pytorch_horovod_benchmark.py , the following error occurred when I reached about 340 steps:
```
what():  Resource temporarily unavailable
*** Process received signal ***
Signal: Aborted (6)
Signal code:  (-6)
/usr/lib64/libpthread.so.0(+0xf5f0)[0x7efe8e2365f0]
/usr/lib64/libc.so.6(gsignal+0x37)[0x7efe8d786337]
/usr/lib64/libc.so.6(abort+0x148)[0x7efe8d787a28]
/usr/local/conda/lib/libstdc++.so.6(_ZN9__gnu_cxx27__verbose_terminate_handlerEv+0xbc)[0x7efe892fe84a]
/usr/local/conda/lib/libstdc++.so.6(+0xabf47)[0x7efe892fcf47]
/usr/local/conda/lib/libstdc++.so.6(+0xabf7d)[0x7efe892fcf7d]
/usr/local/conda/lib/libstdc++.so.6(__cxa_rethrow+0x0)[0x7efe892fd15a]
/usr/local/conda/lib/libstdc++.so.6(_ZSt20__throw_system_errori+0x74)[0x7efe892fa2f3]
/usr/local/conda/lib/libstdc++.so.6(_ZNSt6thread15_M_start_threadESt10shared_ptrINS_10_Impl_baseEEPFvvE+0x91)[0x7efe893194d7]
/usr/local/conda/lib/libstdc++.so.6(_ZNSt6thread15_M_start_threadESt10shared_ptrINS_10_Impl_baseEE+0x7e)[0x7efe8931955e]
/usr/local/lib64/python3.6/site-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(+0x40570)[0x7efc73caa570]
 /usr/local/lib64/python3.6/site-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(+0x49057)[0x7efc73cb3057]
 /usr/local/lib64/python3.6/site-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(+0x4abab)[0x7efc73cb4bab]
/usr/local/conda/lib/libstdc++.so.6(+0xc8421)[0x7efe89319421]
/usr/lib64/libpthread.so.0(+0x7e65)[0x7efe8e22ee65]
/usr/lib64/libc.so.6(clone+0x6d)[0x7efe8d84e88d]
 *** End of error message ***
```
Can you help me see why? I will be very grateful

**Comment by pengyanghua** [comment]
@Richie-yan Could you give more details about your setup as there are many factors affecting the speedup, e.g., NCCL version, pytorch version, horovod version, network bandwidth, GPU type? The setup in the paper is:
1.  Docker image https://github.com/bytedance/byteps/blob/bytescheduler/bytescheduler/docker/pytorch_horovod.Dockerfile
2. NCCL 2.3
3. using 16 - 64 Tesla-V100-PCIE-32GB GPUs
4. TCP
5. python pytorch_horovod_benchmark.py --num-iters 100 --model resnet50 --num-classes 10000
6. In addition, you may tune the partition size, credit size, horovod tensor fusion threshold and horovod cycle time according to Horovod trace, to maximize the overlap of forward and communication.

**Comment by Richie-yan** [comment]
Hi @pengyanghua 
Thank you for your reply. I will try to analyze Horovod's timeline and debug the parameters you mentioned. My environment is as follows:
1. NCCL 2.4.8
2. horovod 0.16.0
3. using 8 Tesla-V100-PCIE-32GB GPUS
4. TCP
I will reply here again if there are new developments

**Comment by Richie-yan** [comment]
> @Richie-yan Could you give more details about your setup as there are many factors affecting the speedup, e.g., NCCL version, pytorch version, horovod version, network bandwidth, GPU type? The setup in the paper is:
> 
> 1. Docker image https://github.com/bytedance/byteps/blob/bytescheduler/bytescheduler/docker/pytorch_horovod.Dockerfile
> 2. NCCL 2.3
> 3. using 16 - 64 Tesla-V100-PCIE-32GB GPUs
> 4. TCP
> 5. python pytorch_horovod_benchmark.py --num-iters 100 --model resnet50 --num-classes 10000
> 6. In addition, according to your testbed setup, you may tune the partition size, credit size, horovod tensor fusion threshold and horovod cycle time according to profiled Horovod timeline, to maximize the overlap of forward and communication. Basically, the horovod cycle time affects the scheduling time of allreduce operations; the tensor fusion threshold by default is 64MB, which should be set to a smaller value, e.g., 16MB; the credit size is typically 2-4 times of the partition size.
> 
> The speedup is different dependent on your testbed. You may profile the baseline using Horovod timeline first, then analyze ideally how much performance gain you can achieve by overlapping communication with forward computation. For ResNet50, if much of the communication is finished during backward, then the improvement space is little. With bytescheduler's overhead, the performance could be worse. In this case, you may set the partition size and credit size to very large values to disable scheduling, and achieve similar performance to the baseline.

**Comment by Richie-yan** [comment]
Make sure in your python pytorch_horovod_benchmark.py --num-iters 100 --model resnet50 --num-classes 10000, is num-classes 1000 or 10000?

**Comment by pengyanghua** [comment]
@Richie-yan  Yes the num of classes is 10000. I did not met the issue you encountered. I guess it is caused by software version, e.g., pytorch version.

**Comment by lucasleesw** [comment]
Hi, My Horovod is faster than Bytescheduler too.
I set HOROVOD_FUSION_THRESHOLD=16MB like mentioned above. And I tried tuned BYTESCHEDULER_CREDIT and BYTESCHEDULER_PARTITION.
The highest number of bytescheduler in my situation is Img/sec per GPU: 110
While horovod is Img/sec per GPU: 130
I use 3 2080ti gpus with vgg16 model and the num-classes is 10000.

Any advice would help, thank you.

**Comment by Richie-yan** [comment]
Hi @pengyanghua 
I created the same mirror image in the same way as your dockfile above, but the following error will still be reported:
<img width="1088" alt="截屏2020-12-18 下午7 55 23" src="https://user-images.githubusercontent.com/41471499/102612253-58ad7180-416b-11eb-8396-a6698a536b98.png">

**Comment by pengyanghua** [comment]
@Richie-yan It seems a mpi issue. You may try Google to solve it. Take a look at https://stackoverflow.com/questions/12079087/fork-retry-resource-temporarily-unavailable.

**Comment by pengyanghua** [comment]
@lucasleesw Can you provide the horovod traces of baseline and bytescheduler respectively and upload their json trace files?

**Comment by Richie-yan** [comment]
Hi, @pengyanghua 
I wrote a very simple model myself and observed the following phenomena:
`BYTESCHEDULER_CREDIT: 16000000 BYTESCHEDULER_PARTITION: 8000000`
Throughput: 40 img/sec per GPU
Timelien's performance is: basically the gradient of the front layer will be scheduled first, and the gradient of the latter layer will be scheduled last
When BYTESCHEDULER_CREDIT is set to: 90000000
Throughput: 56 img/sec per GPU
Timelien's performance is: the gradient of the latter layer will be scheduled first, then the gradient of the previous layer will be scheduled, and finally the remaining gradient of the latter layer will be scheduled
My understanding is that due to the larger gradient of the back layer of my model, the advantages of bytescheduler will be more fully utilized (the front layer gradient will be scheduled first)
But what can’t be explained is: Why does the throughput increase so much after increasing BYTESCHEDULER_CREDIT? Shouldn't the first case be the optimal scheduling?
My model code is as follows:
```
import torch 
import torch.nn as nn
import torch.nn.functional as F

class MyNet(nn.Module):
    def __init__(self):
        super(MyNet, self).__init__()

        self.layer_1 = []  
        for i in range(100):
            self.layer_1.append(nn.Linear(4096, 1)) 
        self.layer_1.append(nn.ReLU(True))
        self.layer_1.append(nn.Linear(1, 4096))
        self.layer_1.append(nn.ReLU(True))                                                                                                                                                                  
    
        self.layer_2 = []  
        for i in range(2):
            self.layer_2.append(nn.Linear(4096, 25088))
        self.layer_2.append(nn.ReLU(True))
        self.layer_2.append(nn.Linear(25088, 4096))
            if i == 1:
            self.layer_2.append(nn.Sigmoid())
            else:
            self.layer_2.append(nn.ReLU(True))
    
    
    self.model_1 = nn.Sequential(*self.layer_1)
    self.model_2 = nn.Sequential(*self.layer_2)

    def forward(self, x): 
        output_1 = self.model_1(x)
        output_2 = self.model_2(output_1)
        return output_2
```
Any advice would help, thank you.

**Comment by lucasleesw** [comment]
@pengyanghua 
Hi, this is timeline of default Horovod for baseline.
[horovod_timeline.log](https://github.com/bytedance/byteps/files/5729516/horovod_timeline.log)
Performance:
Img/sec per GPU: 138.8 +-5.9
Total img/sec on 3 GPU(s): 416.5 +-17.8

And these are timelines for ByteScheduler and Horovod.
[bytescheduler_timeline.log](https://github.com/bytedance/byteps/files/5729549/bytescheduler_timeline.log)
[bytescheduler_horovod_timeline.log](https://github.com/bytedance/byteps/files/5729551/bytescheduler_horovod_timeline.log)
Performance:
Img/sec per GPU: 111.7 +-7.7
Total img/sec on 3 GPU(s): 335.1 +-23.0

Thank you for your help~

**Comment by Richie-yan** [comment]
Hi, @pengyanghua 
By the way, have you tested that when --num-classes is set to 1000, Bytescheduler has a positive return relative to horovod?

**Comment by pengyanghua** [comment]
@Richie-yan Hi.

"But what can’t be explained is: Why does the throughput increase so much after increasing BYTESCHEDULER_CREDIT? Shouldn't the first case be the optimal scheduling?"

> Theoretically the first case should be the optimal scheduling. BYTESCHEDULER_CREDIT represents how much data can be sent out without acknowledgement. The larger BYTESCHEDULER_CREDIT is, the network is more utilized, but with suboptimal 
 tensor communication scheduling. So this is a tradeoff and BYTESCHEDULER_CREDIT depends on your network setup. You can check Sec. 4.2 in the paper for a more clear explanation.

"have you tested that when --num-classes is set to 1000, Bytescheduler has a positive return relative to horovod?"

> No for our 100Gbps RDMA testbed. The positive return may be observed if the network communication is somewhat a bottleneck.

**Comment by pengyanghua** [comment]
@lucasleesw Thanks for uploading the necessary info! I will check the trace this weekend.

**Comment by pengyanghua** [comment]
@lucasleesw From the horovod baseline time log, we can see that there is no opportunity to improve the training performance. 

In your testbed setup with intra-machine PCIe communication, the communication time is fully hidden by backward computation. When layer 0 finished the backward computation, most or all of all-reduce communication of other layers have been finished.  So there is no need to schedule communication in your case.

The worse performance of ByteScheduler comes from tensor partitioning. After partitioning, the total time of all-reduce partitions is larger than the time of all-reduce one big tensor. To see same performance of ByteScheduler as baseline, just disable scheduling by setting BYTESCHEDULER_CREDIT and BYTESCHEDULER_PARTITION to infinity.

To see improvement from bytescheduler, you may try inter-machine communication and do the above analysis to see if there is any speedup.

**Comment by lucasleesw** [comment]
@pengyanghua Thank you for your clearance. I have tested inter-machine scenario and found speedup with vgg16 model. Indeed, schedule communication is necessary or not differ among cases.

**Comment by Richie-yan** [comment]
Hi, @pengyanghua 
Thank you for your answer. I tested resnet50 and VGG16 on 16GPUs and 32GPUs, and got positive benefits. It seems that in the case of multiple nodes, when the communication bottleneck occurs or the communication cannot be fully overlapped by the reverse calculation, ByteScheduler will play effect

**Comment by pengyanghua** [comment]
@Richie-yan Yes, the speedup of ByteScheduler depends on the model forward time, backward time, communication time and scheduling overhead.
